pkg/daemon/daemon_test.go

```diff
@@ -0,0 +1,169 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/libcalico-go/lib/api"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/typha/fv-tests"
+	"github.com/projectcalico/typha/pkg/config"
+	"github.com/projectcalico/typha/pkg/syncclient"
+	"github.com/projectcalico/typha/pkg/syncserver"
+)
+
+var configContents = []byte(`[default]
+LogFilePath=none
+`)
+
+var _ = Describe("Daemon", func() {
+	var d *TyphaDaemon
+	var backend *mockBackend
+	var newClientErr error
+	var earlyLoggingConfigured, loggingConfigured bool
+
+	BeforeEach(func() {
+		d = New()
+		backend = &mockBackend{}
+		d.NewBackendClient = func(config api.CalicoAPIConfig) (c BackendClient, err error) {
+			return backend, newClientErr
+		}
+		earlyLoggingConfigured = false
+		loggingConfigured = false
+		d.ConfigureEarlyLogging = func() {
+			earlyLoggingConfigured = true
+		}
+		d.ConfigureLogging = func(config *config.Config) {
+			Expect(config).ToNot(BeNil())
+			loggingConfigured = true
+		}
+	})
+
+	It("shouldn't panic when DoEarlyRuntimeSetup is called", func() {
+		d.DoEarlyRuntimeSetup()
+		Expect(earlyLoggingConfigured).To(BeTrue())
+	})
+
+	It("should parse the config file path", func() {
+		d.ParseCommandLineArgs([]string{"-c", "/tmp/config.cfg"})
+		Expect(d.ConfigFilePath).To(Equal("/tmp/config.cfg"))
+	})
+
+	It("should default the config file path", func() {
+		d.ParseCommandLineArgs([]string{})
+		Expect(d.ConfigFilePath).To(Equal("/etc/calico/typha.cfg"))
+	})
+
+	Describe("with a config file loaded", func() {
+		var configFile *os.File
+
+		BeforeEach(func() {
+			var err error
+			configFile, err = ioutil.TempFile("", "typha")
+			Expect(err).NotTo(HaveOccurred())
+
+			_, err = configFile.Write(configContents)
+			Expect(err).NotTo(HaveOccurred())
+			err = configFile.Close()
+			Expect(err).NotTo(HaveOccurred())
+
+			d.ParseCommandLineArgs([]string{"-c", configFile.Name()})
+			d.LoadConfiguration()
+			Expect(loggingConfigured).To(BeTrue())
+		})
+		AfterEach(func() {
+			os.Remove(configFile.Name())
+		})
+
+		It("should load the configuration and connect to the datastore", func() {
+			Expect(backend.initCalled).To(BeTrue())
+		})
+
+		It("should create the server components", func() {
+			d.CreateServer()
+			Expect(d.SyncerToValidator).ToNot(BeNil())
+			Expect(d.Syncer).ToNot(BeNil())
+			Expect(d.SyncerToValidator).ToNot(BeNil())
+			Expect(d.ValidatorToCache).ToNot(BeNil())
+			Expect(d.Validator).ToNot(BeNil())
+			Expect(d.Cache).ToNot(BeNil())
+			Expect(d.Server).ToNot(BeNil())
+		})
+
+		It("should start a working server", func() {
+			// Bypass the config validation to tell the server to pick a random port (so we won't clash)
+			d.ConfigParams.ServerPort = syncserver.PortRandom
+			d.CreateServer()
+
+			// Start the server with a context that we can cancel.
+			cxt, cancelFn := context.WithCancel(context.Background())
+			defer cancelFn()
```

Comment:
The context is really there just to let the tests tear it down.  Not sure what I can do to improve on that.